activesupport/lib/active_support/callbacks.rb

```diff
@@ -936,26 +936,33 @@ def define_callbacks(*names)
           options = names.extract_options!
 
           names.each do |name|
+            name_start, name_end =
+              if (md = name.match(/[!\?]\z/))
```

Comment:
consider making it possible to define callbacks on methods which end in `!` or `?`
included in this PR since it makes the callback definition simpler, if that is an acceptable route to follow

Reply:
I don't think there's any advantage to naming the callback with the `!` -- it's in the name of a private method that's involved, but I don't think that's a user-facing concern.

(Off the top of my head, I think I might actually favour `schema` or `schema_load` over `load_schema` as a callback name here -- though that's a weak thought, without considering existing naming precedents.)

Reply:
@matthewd Thanks for taking a look.  So, you the think the ideal, practical implementation should be in lib/active_record/model_schema.rb

```diff
        def load_schema
          return if schema_loaded?
          @load_schema_monitor.synchronize do
            return if defined?(@columns_hash) && @columns_hash

-            load_schema!
+            run_callbacks :load_schema do
+              load_schema!
+            end

            @schema_loaded = true
          rescue
            reload_schema_from_cache # If the schema loading failed half way through, we must reset the state.
            raise
          end
        end
```

Reply:
The reason why I tried adding the callbacks in `activerecord/lib/active_record/callbacks.rb` via `define_model_callbacks :load_schema!, only: :after` is because Callbacks are include into ActiveRecord::Base after ModelSchema

Reply:
I think you're expecting `define_model_callbacks` to do more than it does?

Reply:
responding to https://github.com/rails/rails/pull/46205#issuecomment-1283135279 here since it lets us keep the topic in a thread

@matthewd Perhaps we can find some time to chat? I think this callback needs some special handling because it is acting on the record class, on the record instance.  All the callbacks defined by `define_model_callbacks` would be used on the instance like `def create_or_update(**)  _run_save_callbacks { super } end` but to be able to write `_run_load_schema_callbacks { super }` you'd need to do something like extend the ClassMethods, in addition to the existing includes

I'd really like to rely on the ActiveSupport::Callbacks code and anything that builds on it, of course, since anything else would be surprising, but perhaps we can agree on some MVP interface for where the `_run_load_schema_callbacks { super }` code or equivalent goes and work backwards from there?